instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/HostIdResourceProvider.java

```diff
@@ -0,0 +1,176 @@
+      this.exitCode = exitCode;
+      this.lines = lines;
+    }
+  }
+
+  public HostIdResourceProvider() {
+    this(
+        HostIdResourceProvider::getOsType,
+        path -> {
+          try {
+            return Files.readAllLines(path);
+          } catch (IOException e) {
+            throw new IllegalStateException(e);
+          }
+        },
+        HostIdResourceProvider::queryWindowsRegistry);
+  }
+
+  // Visible for testing
+  HostIdResourceProvider(
+      Supplier<OsType> getOsType,
+      Function<Path, List<String>> pathReader,
+      Supplier<ExecResult> queryWindowsRegistry) {
+    this.getOsType = getOsType;
+    this.pathReader = pathReader;
+    this.queryWindowsRegistry = queryWindowsRegistry;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    OsType osType = getOsType.get();
+    switch (osType) {
+      case WINDOWS:
+        return readWindowsGuid();
+      case LINUX:
+        return readLinuxMachineId();
+    }
+    throw new IllegalStateException("Unsupported OS type: " + osType);
+  }
+
+  private Resource readLinuxMachineId() {
+    Path path = FileSystems.getDefault().getPath("/etc/machine-id");
+    try {
+      List<String> lines = pathReader.apply(path);
+      if (lines.isEmpty()) {
+        logger.warning("Failed to read /etc/machine-id: empty file");
+        return Resource.empty();
+      }
+      return Resource.create(Attributes.of(HOST_ID, lines.get(0)));
+    } catch (RuntimeException e) {
+      logger.log(Level.WARNING, "Failed to read /etc/machine-id", e);
+      return Resource.empty();
+    }
+  }
+
+  private static OsType getOsType() {
+    String osName = System.getProperty("os.name");
+    return osName != null && osName.startsWith("Windows") ? OsType.WINDOWS : OsType.LINUX;
+  }
+
+  private Resource readWindowsGuid() {
+
+    try {
+      ExecResult execResult = queryWindowsRegistry.get();
+
+      if (execResult.exitCode != 0) {
+        logger.warning(
+            "Failed to read Windows registry. Exit code: "
+                + execResult.exitCode
+                + " Output: "
+                + String.join("\n", execResult.lines));
+        return Resource.empty();
+      }
+
+      for (String line : execResult.lines) {
+        if (line.contains("MachineGuid")) {
+          String[] parts = line.trim().split("\\s+");
+          if (parts.length == 3) {
+            return Resource.create(Attributes.of(HOST_ID, parts[2]));
+          }
+        }
+      }
+      logger.warning(
+          "Failed to read Windows registry: No MachineGuid found in output: " + execResult.lines);
+      return Resource.empty();
+    } catch (RuntimeException e) {
+      logger.log(Level.WARNING, "Failed to read Windows registry", e);
+      return Resource.empty();
+    }
+  }
+
+  private static ExecResult queryWindowsRegistry() {
+    try {
+      Process process = Runtime.getRuntime().exec(REGISTRY_QUERY);
```

Comment:
good point!
I think a thread is overkill, but I added a check - wait for 2 seconds and then destroy the process. This will also handle the case where it's waiting for the output to be consumed.